src/architecture/storage.md

```diff
@@ -1,262 +1,158 @@
 
-[`ProofMapIndex`][proof-map-index] is a Merkelized version of a map
-based on Merkle Patricia tree.
-It implements the same methods as the `MapIndex`, adding the ability to
-create proofs of existence for its key-value pairs, or proofs of absence
-if a key is absent in the map. The following additional
-procedures are supported:
+- **list**, where either keys or values are stored
+- **map**, where key-value pairs are stored
+- **set** of unique items
+- **entry**, which represents a set containing only one item.
 
-- Get the root node’s value
-- Build a proof for the requested key. Tree proves either key
-  existence (and its value), or key absence
+and also merkelized versions of lists and maps.
 
-## Low-level Storage
+Read more about MerkleDB [here][merkledb].
 
-Exonum uses third-party database engines to persist blockchain state
-locally. To use the particular database, a minimal [`Database`][database]
-interface should be implemented for it:
+1. [System Root Objects](#system-root-objects) lists the types of data
+   stores supported by Exonum. Objects represent the highest abstraction level
+   for the data storage.
+2. [Low-level storage](#low-level-storage) explains how objects are persisted
+   using RocksDB.
+3. [View layer](#view-layer) describes the wrapper over the DB engine
+   that ensures atomicity of blocks and transactions.
+4. [Indexing](#indexing) explains how indices over structured data
+   can be built in Exonum.
 
-- Get a value by a [column family][col-family] name and a key
-- Put a new value at the specified column family / key (insert or update
-  the saved one)
-- Delete a key-value pair by column family name / key
+## System Root Objects
 
-All the tables functionality is reduced to these atomic call types.
+All [root objects][blockchain-schema] in Exonum have `core.*` prefix. Below is a
+list of available objects that are used in client applications.
 
-Currently the main database engine is [RocksDB][rocks-db].
+- `transactions: MapIndex`  
+  Represents a map from the transaction hash into a raw transaction structure.
+- `transaction_results: ProofMapIndex`  
+  Keeps execution results for all accepted transactions,
+  indexed by transaction hashes.
+- `transactions_pool: KeySetIndex`  
+  Stores the set of hashes of the known transactions that have not been
+  committed yet.
+- `transactions_pool_len: Entry`  
+  Caches the number of entries in `transactions_pool`.
+- `transactions_locations: MapIndex`  
+  For every transaction hash keeps the position of said transaction inside the
+  block and the block height.
+- `blocks: MapIndex`  
+  Stores the block object for every block height.
+- `block_hashes_by_height: ListIndex`  
+  Saves block hashes indexed by block heights.
+- `block_transactions: ProofListIndex`  
+  Stores a list of transactions of the specific block indexed by the block
+  height.
+- `precommits: ListIndex`  
+  Stores a list of precommits of the validators for the specific block, indexed
```

Comment:
> indexed by the block hash.

That sounds confusing without any introduction to families (is that still a thing?). How a _list_ is indexed by a _hash_?

Reply:
edited